examples/charm++/cuda/direct/jacobi2d/jacobi2d.C

```diff
@@ -0,0 +1,409 @@
+    hapiCheck(cudaMallocHost((void**)&h_right_ghost, sizeof(double) * block_size));
+    hapiCheck(cudaMallocHost((void**)&h_bottom_ghost, sizeof(double) * block_size));
+    hapiCheck(cudaMallocHost((void**)&h_top_ghost, sizeof(double) * block_size));
+    hapiCheck(cudaMalloc((void**)&d_left_ghost, sizeof(double) * block_size));
+    hapiCheck(cudaMalloc((void**)&d_right_ghost, sizeof(double) * block_size));
+    cudaStreamCreate(&stream);
+
+    // Initialize temperature data
+    invokeInitKernel(d_temperature, block_size, stream);
+
+    CkCallback cb(CkReductionTarget(Main, initDone), main_proxy);
+    contribute(cb);
+  }
+
+  void sendGhosts(void) {
+    // Pack non-contiguous ghosts to temporary contiguous buffers on device
+    invokePackingKernels(d_temperature, d_left_ghost, d_right_ghost, left_bound,
+        right_bound, block_size, stream);
+
+    if (use_zerocopy) {
+      // Send ghosts to neighboring chares
+      if (!left_bound)
+        thisProxy(x - 1, y).receiveGhostsZC(my_iter, RIGHT, block_size,
+            CkDeviceBuffer(d_left_ghost, stream));
+      if (!right_bound)
+        thisProxy(x + 1, y).receiveGhostsZC(my_iter, LEFT, block_size,
+            CkDeviceBuffer(d_right_ghost, stream));
+      if (!top_bound)
+        thisProxy(x, y - 1).receiveGhostsZC(my_iter, BOTTOM, block_size,
+            CkDeviceBuffer(d_temperature + (block_size + 2) + 1, stream));
+      if (!bottom_bound)
+        thisProxy(x, y + 1).receiveGhostsZC(my_iter, TOP, block_size,
+            CkDeviceBuffer(d_temperature + (block_size + 2) * block_size + 1, stream));
+    }
+    else {
+      // Transfer ghosts from device to host
+      hapiCheck(cudaMemcpyAsync(h_left_ghost, d_left_ghost, block_size * sizeof(double),
+            cudaMemcpyDeviceToHost, stream));
+      hapiCheck(cudaMemcpyAsync(h_right_ghost, d_right_ghost, block_size * sizeof(double),
+            cudaMemcpyDeviceToHost, stream));
+      hapiCheck(cudaMemcpyAsync(h_top_ghost, d_temperature + (block_size + 2) + 1,
+            block_size * sizeof(double), cudaMemcpyDeviceToHost, stream));
+      hapiCheck(cudaMemcpyAsync(h_bottom_ghost, d_temperature + (block_size + 2) * block_size + 1,
+            block_size * sizeof(double), cudaMemcpyDeviceToHost, stream));
+      cudaStreamSynchronize(stream);
+
+      // Send ghosts to neighboring chares
+      if (!left_bound)
+        thisProxy(x - 1, y).receiveGhostsReg(my_iter, RIGHT, block_size, h_left_ghost);
+      if (!right_bound)
+        thisProxy(x + 1, y).receiveGhostsReg(my_iter, LEFT, block_size, h_right_ghost);
+      if (!top_bound)
+        thisProxy(x, y - 1).receiveGhostsReg(my_iter, BOTTOM, block_size, h_top_ghost);
+      if (!bottom_bound)
+        thisProxy(x, y + 1).receiveGhostsReg(my_iter, TOP, block_size, h_bottom_ghost);
+    }
+  }
+
+  void receiveGhostsZC(int ref, int dir, int &w, double *&buf, CkDeviceBufferPost *devicePost) {
```

Comment:
Since this will be an example, worth adding a comment here saying this is the "Post Entry method and the matching regular entry method is in the .ci file"

Reply:
👍